### ISSUE TITLE
Auto-assignment

### ISSUE THREAD

**Issue by XAMPPRocky** [rust-lang/triagebot#558]
Highfive currently auto-assigns people to review PRs based (I believe) on some match between the directory where the changes occur and its JSON config containing assignees. It would be great if we could not implement auto-assignment in triagebot, but also make some improvements to the auto-assignment system.

### Requirements
- Instead of using a JSON config, assignees should taken from data in `rust-lang/team`.

### Potential Improvements
- On [zulip there was a discussion][z] around how to improve review latency, or at least the review experience. One proposed idea was place a limit how many open `S-waiting-on-review` PRs you as a rust-lang member could be assigned, and if all members of a team are assigned a new `S-waiting-on-reviewer` is given to the PR. Which would also allow us to properly communicate a team's current bandwidth. 
#### Unresolved Questions
  - How would people set their limit?
    - It would be easiest (I would guess) on the implementation side if this limit was also set in `team`, but team can take a some time to get a PR merged.
    - Because of the above I feel it would be better suited as a bot command on Zulip.
    - The solution could also be both, a user sets through a bot command which updates the team repo.

[z]: https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.22Review.20latency.20can.20be.20challenging.22

**Comment by dtolnay** [comment]
You suggested in #triage-wg that you would be interested in feedback from a libs team perspective so I'll weigh in here.<br><br>

> if all members of a team are assigned, a new `S-waiting-on-reviewer` is given to the PR.

I think about this problem space in terms of throughput. Simplifying a bit, triage wg is responsible over a pool of open PRs. This pool has an inflow (PRs being opened) at some rate, and an outflow (completing the review and merging or closing) at some rate.

- Examples that contribute to increase inflow: building an inclusive and friendly community; improving the contribution process by e.g. documentation like the rustc dev guide; exponential scaling of the Rust user base.
- Examples that contribute to increase outflow: productivity improvements via automation (thank you infra team); building organizational experience with scaling and structuring teams better; recommendations to prioritize reviews over other team business as the compiler team has been doing; entrusting parts of the process to nonmembers of the team.

Within this terminology, the suggestion of <kbd>S-waiting-on-reviewer</kbd> pertains to what should be done if the inflow temporarily spikes above the outflow for some team.

A team would generally identify with one of three regimes: inflow = outflow ("running smoothly"), inflow spiked above outflow but they'll catch up soon enough ("spikes"), or sustained inflow &gt; outflow ("backlogged").

Without giving specific dates, I think the libs team has been 1 - 1.5 years ahead of other teams in transitioning from "running smoothly" to "spikes" to "backlogged". Note that libs team composition is a bit unlike other teams in that it's typical to have active library projects outside the standard library that take a majority of a member's time, which is unlike the compiler team or lang team where the team role is usually the dominant project.

So from my perspective addressing the "spikes" regime seems well motivated because it's where many teams already are or will soon be, but I encourage starting to think about the next step as well. This is probably a bit beyond this thread, but the relevant question would be: Is it going to continue to be reasonable that anyone can send a PR at any time and it will have the team's attention? if yes, how, and if not, what is triage wg's role in helping teams make the best use of time under that regime and in maintaining a positive experience for contributors.

**Comment by pietroalbini** [comment]
I don't think taking assignees directly from the team repo is a good choice: I want changing them to be as frictionless as possible, and the team repo would get messy if we had per-directory permissions for each repo that wants auto-assignments.

An approach I'd prefer is to use [`CODEOWNERS`](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) files, and mandate that only teams managed through the team repo should be assigned instead of individuals.

**Comment by XAMPPRocky** [comment]
> and the team repo would get messy if we had per-directory permissions for each repo that wants auto-assignments.

Sorry I should clarify what I meant, which is that only the member data, such as who is in what team would taken from `rust-lang/team`. The actual "codeowners" part of what glob path triggers which team/person to ping could be set in the repository's `triagebot.toml`, or it could be its own file in the repo.

> An approach I'd prefer is to use CODEOWNERS files, and mandate that only teams managed through the team repo should be assigned instead of individuals.

I think assigning teams over members would make the bandwidth problem worse, because as I understand that would mean that every member gets pinged each time.

**Comment by pietroalbini** [comment]
> I think assigning teams over members would make the bandwidth problem worse, because as I understand that would mean that every member gets pinged each time.

Well we could have triagebot resolve team members and still assign an individual.

**Comment by XAMPPRocky** [comment]
That is what I was thinking as well, owners set to teams and then triagebot figures out who amongst them is available to review.

I was also wrong on how the team code owner reviews on GitHub works, it does allow you to assign to individuals from a team, and choose how to distribute the load, which is most of what we want I think. It doesn’t have a way to limit how many reviews though currently. https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/managing-code-review-assignment-for-your-team

**Comment by ehuss** [comment]
With https://github.com/rust-lang/triagebot/pull/1656, I think this issue can be closed.

If someone wants a more sophisticated load-balancing system, I think it would be good to open a separate issue for that with a specific proposal. There are some notes [here](https://github.com/rust-lang/triagebot/blob/d7c405b8c1067a94300c943cb4070de238b285c4/src/handlers/assign.rs#L582-L600) about some different options (GitHub has some advanced options for example). However, from my perspective I feel like load balancing may not work very well. It's not unusual for PRs to be left open intentionally in a state where it isn't exactly waiting on the assignee to do something, but also not exactly waiting on the author, either. I think it would be tricky to get something that makes a significant improvement and is also not very complex.

The random-choice algorithm is very primitive, and I think can be unfair, but I assume if someone is feeling overwhelmed, they can reroll. Or the triage team can help reroll if necessary.
